tests/dismod/db/test_wrapper.py

```diff
@@ -189,42 +189,85 @@ def test_validate_data__missing_column():
-        engine = _get_engine(None)
-        dm_file = DismodFile(engine, column_set, column_set)
-        dm_file.integrand = pd.DataFrame(
-            {"integrand_id": [0], "integrand_name": ["Sincidence"], "minimum_meas_cv": [0.0]}
-        )
-        dm_file.density = pd.DataFrame({"density_id": [0], "density_name": ["uniform"]})
-        dm_file.node = pd.DataFrame({"node_id": [0], "node_name": ["Foreverland"], "parent": [np.NaN]})
-        dm_file.weight = pd.DataFrame({"weight_id": [0], "weight_name": ["uniform"], "n_age": [1], "n_time": [1]})
-        common_input = {
-            "data_id": [0],
-            "data_name": ["only"],
-            "integrand_id": [0],
-            "density_id": [0],
-            "node_id": [0],
-            "weight_id": [0],
-            "hold_out": [0],
-            "meas_value": [1.0],
-            "meas_std": [0.1],
-            "eta": [0.0],
-            "nu": [0.0],
-            "age_lower": [0.0],
-            "age_upper": [10.0],
-            "time_lower": [1971],
-            "time_upper": [2020],
-        }
-        for col_title, col_type in column_set.items():
-            common_input[col_title] = col_type()
+def test_write_covariate_column__success(base_file):
+    new_data = pd.DataFrame(
+        {
+            "data_name": "foo",
+            "integrand_id": 1,
+            "density_id": 1,
+            "node_id": 1,
+            "weight_id": 1,
+            "hold_out": 0,
+            "meas_value": 0.0,
+            "meas_std": 0.0,
+            "eta": np.nan,
+            "nu": np.nan,
+            "age_lower": 0,
+            "age_upper": 10,
+            "time_lower": 1990,
+            "time_upper": 2000,
+            "x_s_source": 0,
+            "x_sex": 2.0,
+        },
+        index=[0],
+    )
+    base_file.data = new_data
+    base_file.flush()
 
-        dm_file.data = pd.DataFrame(common_input)
-        dm_file.flush()
-        dm_files.append(dm_file)
 
-    for check_set, in_file in zip(column_sets, dm_files):
-        for check_col in check_set:
-            assert check_col in in_file.data.columns
+def test_read_covariate_column__success(base_file):
+    new_data = pd.DataFrame(
+        {
+            "data_name": "foo",
+            "integrand_id": 1,
+            "density_id": 1,
+            "node_id": 1,
+            "weight_id": 1,
+            "hold_out": 0,
+            "meas_value": 0.0,
+            "meas_std": 0.0,
+            "eta": np.nan,
+            "nu": np.nan,
+            "age_lower": 0,
+            "age_upper": 10,
+            "time_lower": 1990,
+            "time_upper": 2000,
+            "x_s_source": 0,
+            "x_sex": 2.0,
+        },
+        index=[0],
+    )
+    base_file.data = new_data
+    base_file.flush()
+
+    new_dm = DismodFile(base_file.engine)
+    assert list(new_dm.data.x_s_source) == [0]
+    assert list(new_dm.data.x_sex) == [2.0]
+
+
+def test_write_covariate_column__bad_name(base_file):
+    new_data = pd.DataFrame(
```

Comment:
Looks like this one is the same as above.  Maybe share it?

Reply:
Valid